mocpy/moc.py

```diff
@@ -2,427 +2,360 @@
+        Add all the pixels at max order in the neighbourhood of the moc
+
         """
-        return self._query(table_id, max_rows)
+        from astropy_healpix import HEALPix
 
-    def _query(self, resource_id, max_rows):
+        hp = HEALPix(nside=(1 << self.max_order), order='nested')
+
+        pix_arr = np.array(list(self.best_res_pixels_iterator()))
+        neighbour_pix_arr = AbstractMoc._neighbour_pixels(hp, pix_arr)
+
+        neighbour_pix_arr = np.setdiff1d(neighbour_pix_arr, pix_arr)
+
+        factor = 4 ** (self.HPY_MAX_NORDER - self.max_order)
+        for pix in neighbour_pix_arr:
+            self._interval_set.add((pix * factor, (pix + 1) * factor))
+
+    def remove_neighbours(self):
         """
-        internal method to query Simbad or a VizieR table
-        for sources in the coverage of the MOC instance
+        Remove all the pixels at max order located at the bound of the moc
+
         """
-        import requests
-        import tempfile, os
+        from astropy_healpix import HEALPix
 
-        if max_rows is not None and max_rows>=0:
-            max_rows_str = str(max_rows)
-        else:
-            max_rows_str = str(9999999999)
+        hp = HEALPix(nside=(1 << self.max_order), order='nested')
 
-        tmp_moc = tempfile.NamedTemporaryFile(delete = False)
-        self.write(tmp_moc.name)
-        r = requests.post('http://cdsxmatch.u-strasbg.fr/QueryCat/QueryCat', data={'mode': 'mocfile' , 'catName': resource_id, 'format': 'votable', 'limit': max_rows_str}, files={'moc': open(tmp_moc.name, 'rb')}, headers={'User-Agent': 'MOCPy'}, stream=True)
-        
-        tmp_vot = tempfile.NamedTemporaryFile(delete = False)
-        with open(tmp_vot.name, 'w') as h:
-            for line in r.iter_lines():
-                if line:
-                    h.write(line.decode(r.encoding)+'\n')
+        pix_arr = np.array(list(self.best_res_pixels_iterator()))
 
-        from astropy.io.votable import parse_single_table
-        table = parse_single_table(tmp_vot.name).to_table()
+        neighbour_pix_arr = AbstractMoc._neighbour_pixels(hp, pix_arr)
 
-        # finally delete temp files
-        os.unlink(tmp_moc.name)
-        os.unlink(tmp_vot.name)
+        only_neighbour_arr = np.setxor1d(neighbour_pix_arr, pix_arr)
 
-        return table
+        bound_pix_arr = AbstractMoc._neighbour_pixels(hp, only_neighbour_arr)
 
-    @classmethod
-    def from_file(cls, local_path, moc_order=None):
-        return MOC_io.read_local(path=local_path, moc_order=moc_order)
+        result_pix_arr = np.setdiff1d(pix_arr, bound_pix_arr)
+
+        factor = 4 ** (self.HPY_MAX_NORDER - self.max_order)
+        self._interval_set.clear()
+        for pix in result_pix_arr:
+            self._interval_set.add((pix * factor, (pix + 1) * factor))
 
     @classmethod
-    def from_vizier_table(cls, table_id, nside=256):
+    def from_image(cls, header, hdu, moc_order, mask_arr=None):
```

Comment:
The parameter `hdu` isn't documented in the docstring.
It is accessed once in the function.

Why make an API where you support both `header, mask_arr` and `hdu`?
Suggest to choose one or the other, but not to try and support both unless it's really important.
Otherwise it's easy to get bugs in some cases, or generally it's harder to learn the API for users and choose.

Reply:
I removed the ``hdu`` param and keep the ``header``, ``mask_arr``implementation only. The user is free to create a mask_arr from the image data array and call the ``from_image(cls, header, mask_arr)`` method.

Reply:
👍